<a href="https://colab.research.google.com/github/Warren-Mbekezeli-Mnisi/Logistics_Regression/blob/main/SENDY_LOGISTICS_MODELS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **SENDY LOGISTICS CHALLENGE**

## **DATA PREPROCESSING**

## Import Libraries 

In [1]:
#import all the important libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

## Import Dataset

### Import the Training DataSet

In [2]:
#import training dataset and observe the results
train_set = pd.read_csv('Train.csv')

y_train = train_set.iloc[:, -1].values

train_set.drop(['User Id', 'Vehicle Type','Order No', 'Arrival at Destination - Weekday (Mo = 1)',
             'Arrival at Destination - Time', 'Arrival at Destination - Day of Month', 'Time from Pickup to Arrival' ],
             axis=1, inplace=True)

test_set = pd.read_csv('Test.csv')
test_set.drop(['User Id', 'Vehicle Type','Order No'  ],
             axis=1, inplace=True)

rider_info = pd.read_csv('Riders.csv')

print(len(train_set.columns))
print(train_set.columns)

print(len(test_set.columns))
print(test_set.columns)

for column in train_set.columns:
  if column not in test_set.columns:
    print(column)

FileNotFoundError: [Errno 2] File Riders.csv does not exist: 'Riders.csv'

## Preproccessing the Datasets

In [28]:
df_train.info() #get detailed information about the train dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21201 entries, 0 to 21200
Data columns (total 29 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   Order No                                   21201 non-null  object 
 1   User Id                                    21201 non-null  object 
 2   Vehicle Type                               21201 non-null  object 
 3   Platform Type                              21201 non-null  int64  
 4   Personal or Business                       21201 non-null  object 
 5   Placement - Day of Month                   21201 non-null  int64  
 6   Placement - Weekday (Mo = 1)               21201 non-null  int64  
 7   Placement - Time                           21201 non-null  object 
 8   Confirmation - Day of Month                21201 non-null  int64  
 9   Confirmation - Weekday (Mo = 1)            21201 non-null  int64  
 10  Confirmation - Time   

Inspect the train dataset to investigate the following:

1. The number of columns and rows
2. The datatype for each column
3. To check for any missing values on each column

from the above .info() method we can conclude that the number of columns is 29, the number of rows(entries) is 21201, we also have different datatypes for different columns, finally it is evident that we have missing values in two columns which are Temperature and Precipitation in millimeters 

In [29]:
df_test.info() #get detailed information about the test dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7068 entries, 0 to 7067
Data columns (total 25 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   Order No                              7068 non-null   object 
 1   User Id                               7068 non-null   object 
 2   Vehicle Type                          7068 non-null   object 
 3   Platform Type                         7068 non-null   int64  
 4   Personal or Business                  7068 non-null   object 
 5   Placement - Day of Month              7068 non-null   int64  
 6   Placement - Weekday (Mo = 1)          7068 non-null   int64  
 7   Placement - Time                      7068 non-null   object 
 8   Confirmation - Day of Month           7068 non-null   int64  
 9   Confirmation - Weekday (Mo = 1)       7068 non-null   int64  
 10  Confirmation - Time                   7068 non-null   object 
 11  Arrival at Pickup

from the above .info() method we can conclude that the number of columns is 25, the number of rows(entries) is 7068, we also have different datatypes for different columns, finally it is evident that we have missing values in two columns which are Temperature and Precipitation in millimeters.

We now have inspected our two dataframes, we can conclude that we need to do data engineering on both dataframes.

The first step is to handle missing entries on both our dataset, as mentioned before the columns in question here are Temperature and Precipitation in millimeters. it will be wise for us to inspect the columns to identify thier datatypes in other for us to be able to handle those missing values

In [30]:
#inspecting the missing values from both the df_train and df_train dataframe
df_train[['Temperature','Precipitation in millimeters']].dtypes

Temperature                     float64
Precipitation in millimeters    float64
dtype: object

In [31]:
#inspecting the missing values from both the df_train and df_train dataframe
df_test[['Temperature','Precipitation in millimeters']].dtypes

Temperature                     float64
Precipitation in millimeters    float64
dtype: object

Now that we are aware that both columns have a float64 datatype we can now decide that we will fill the missing values with the average/mean of each column

In [32]:
#filling the missing values on both the dataset
df_train['Temperature'].fillna((df_train['Temperature'].mean()), inplace=True)
df_train['Precipitation in millimeters'].fillna((df_train['Precipitation in millimeters'].mean()), inplace=True)

df_test['Temperature'].fillna((df_test['Temperature'].mean()), inplace=True)

df_test['Precipitation in millimeters'].fillna((df_test['Precipitation in millimeters'].mean()), inplace=True)

In [33]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21201 entries, 0 to 21200
Data columns (total 29 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   Order No                                   21201 non-null  object 
 1   User Id                                    21201 non-null  object 
 2   Vehicle Type                               21201 non-null  object 
 3   Platform Type                              21201 non-null  int64  
 4   Personal or Business                       21201 non-null  object 
 5   Placement - Day of Month                   21201 non-null  int64  
 6   Placement - Weekday (Mo = 1)               21201 non-null  int64  
 7   Placement - Time                           21201 non-null  object 
 8   Confirmation - Day of Month                21201 non-null  int64  
 9   Confirmation - Weekday (Mo = 1)            21201 non-null  int64  
 10  Confirmation - Time   

From informtion above we can see that our train dataframe we have additional Idependent variables, which results in our dataframe have more (idv) as opposed to the test dataframe, so to fix this we will have to squeeze the ivd of the train dataframe to match with the test (ivd)

In [34]:
new_cols =df_test.columns # assign a new variable to assess the test columns
new_data =df_train[new_cols] # transform the df_train.columns to match excatly with the df_test.columns 
y=np.array(df_train['Time from Pickup to Arrival']).reshape(-1,1) # defining the dependent variable

In [35]:
# inspect the new dataframe
new_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21201 entries, 0 to 21200
Data columns (total 25 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   Order No                              21201 non-null  object 
 1   User Id                               21201 non-null  object 
 2   Vehicle Type                          21201 non-null  object 
 3   Platform Type                         21201 non-null  int64  
 4   Personal or Business                  21201 non-null  object 
 5   Placement - Day of Month              21201 non-null  int64  
 6   Placement - Weekday (Mo = 1)          21201 non-null  int64  
 7   Placement - Time                      21201 non-null  object 
 8   Confirmation - Day of Month           21201 non-null  int64  
 9   Confirmation - Weekday (Mo = 1)       21201 non-null  int64  
 10  Confirmation - Time                   21201 non-null  object 
 11  Arrival at Pick

In [36]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7068 entries, 0 to 7067
Data columns (total 25 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   Order No                              7068 non-null   object 
 1   User Id                               7068 non-null   object 
 2   Vehicle Type                          7068 non-null   object 
 3   Platform Type                         7068 non-null   int64  
 4   Personal or Business                  7068 non-null   object 
 5   Placement - Day of Month              7068 non-null   int64  
 6   Placement - Weekday (Mo = 1)          7068 non-null   int64  
 7   Placement - Time                      7068 non-null   object 
 8   Confirmation - Day of Month           7068 non-null   int64  
 9   Confirmation - Weekday (Mo = 1)       7068 non-null   int64  
 10  Confirmation - Time                   7068 non-null   object 
 11  Arrival at Pickup

Now we have transformed the train dataset to match the testing dataframe columns and assigned it a new variable new_data. Since we are going to use both the dataframe to train,validate and test our model we would want to have concatenate both the dataframes into a single dataframe 

In [37]:
df=pd.concat([new_data,df_test])
df.head()

Order No       User Id Vehicle Type  Platform Type  \
0   Order_No_4211   User_Id_633         Bike              3   
1  Order_No_25375  User_Id_2285         Bike              3   
2   Order_No_1899   User_Id_265         Bike              3   
3   Order_No_9336  User_Id_1402         Bike              3   
4  Order_No_27883  User_Id_1737         Bike              1   

  Personal or Business  Placement - Day of Month  \
0             Business                         9   
1             Personal                        12   
2             Business                        30   
3             Business                        15   
4             Personal                        13   

   Placement - Weekday (Mo = 1) Placement - Time  Confirmation - Day of Month  \
0                             5       9:35:46 AM                            9   
1                             5      11:16:16 AM                           12   
2                             2      12:39:25 PM                           30   
3                             5       9:25:34 AM                           15   
4                             1       9:55:18 AM                           13   

   Confirmation - Weekday (Mo = 1)  ... Pickup - Weekday (Mo = 1)  \
0                                5  ...                         5   
1                                5  ...                         5   
2                                2  ...                         2   
3                                5  ...                         5   
4                                1  ...                         1   

   Pickup - Time  Distance (KM) Temperature  Precipitation in millimeters  \
0    10:27:30 AM              4   20.400000                      7.905797   
1    11:44:09 AM             16   26.400000                      7.905797   
2    12:53:03 PM              3   23.258889                      7.905797   
3     9:43:06 AM              9   19.200000                      7.905797   
4    10:05:23 AM              9   15.400000                      7.905797   

   Pickup Lat Pickup Long  Destination Lat  Destination Long      Rider Id  
0   -1.317755   36.830370        -1.300406         36.829741  Rider_Id_432  
1   -1.351453   36.899315        -1.295004         36.814358  Rider_Id_856  
2   -1.308284   36.843419        -1.300921         36.828195  Rider_Id_155  
3   -1.281301   36.832396        -1.257147         36.795063  Rider_Id_855  
4   -1.266597   36.792118        -1.295041         36.809817  Rider_Id_770  

[5 rows x 25 columns]

We want to change the naming convention of the dataframe to be consistent throught the dataframe

In [38]:
df.columns = [col.replace(" ","_") for col in df.columns] # replace the whitespace on the columns with an underscore 
df.head()

Order_No       User_Id Vehicle_Type  Platform_Type  \
0   Order_No_4211   User_Id_633         Bike              3   
1  Order_No_25375  User_Id_2285         Bike              3   
2   Order_No_1899   User_Id_265         Bike              3   
3   Order_No_9336  User_Id_1402         Bike              3   
4  Order_No_27883  User_Id_1737         Bike              1   

  Personal_or_Business  Placement_-_Day_of_Month  \
0             Business                         9   
1             Personal                        12   
2             Business                        30   
3             Business                        15   
4             Personal                        13   

   Placement_-_Weekday_(Mo_=_1) Placement_-_Time  Confirmation_-_Day_of_Month  \
0                             5       9:35:46 AM                            9   
1                             5      11:16:16 AM                           12   
2                             2      12:39:25 PM                           30   
3                             5       9:25:34 AM                           15   
4                             1       9:55:18 AM                           13   

   Confirmation_-_Weekday_(Mo_=_1)  ... Pickup_-_Weekday_(Mo_=_1)  \
0                                5  ...                         5   
1                                5  ...                         5   
2                                2  ...                         2   
3                                5  ...                         5   
4                                1  ...                         1   

   Pickup_-_Time  Distance_(KM) Temperature  Precipitation_in_millimeters  \
0    10:27:30 AM              4   20.400000                      7.905797   
1    11:44:09 AM             16   26.400000                      7.905797   
2    12:53:03 PM              3   23.258889                      7.905797   
3     9:43:06 AM              9   19.200000                      7.905797   
4    10:05:23 AM              9   15.400000                      7.905797   

   Pickup_Lat Pickup_Long  Destination_Lat  Destination_Long      Rider_Id  
0   -1.317755   36.830370        -1.300406         36.829741  Rider_Id_432  
1   -1.351453   36.899315        -1.295004         36.814358  Rider_Id_856  
2   -1.308284   36.843419        -1.300921         36.828195  Rider_Id_155  
3   -1.281301   36.832396        -1.257147         36.795063  Rider_Id_855  
4   -1.266597   36.792118        -1.295041         36.809817  Rider_Id_770  

[5 rows x 25 columns]

## EDA

## **MODEL #1**

## **MODEL #2**

## **MODEL #3**